subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java

```diff
@@ -131,96 +124,58 @@ private void resizeTo(int newBuildProgressLabelCount) {
-        if (!unusedProgressLabels.isEmpty()) {
-            association = new AssociationLabel(operation, unusedProgressLabels.pop());
+        // No more unused label? Try to resize.
+        if (unusedProgressLabels.isEmpty()) {
+            int newValue = operationIdToAssignedLabels.size() + 1;
+            resizeTo(newValue);
         }
 
-        if (association == null) {
-            unassignedProgressOperations.addLast(operation);
+        // Try to use a new label
+        if (unusedProgressLabels.isEmpty()) {
+            unassignedProgressOperations.add(operation);
         } else {
-            operationIdToAssignedLabels.put(operation.getOperationId(), association);
+            attach(operation, unusedProgressLabels.pop());
         }
     }
 
-    private void detach(ProgressOperation operation) {
-        if (operation.getParent() != null) {
-            removeDirectChildOperationId(operation.getParent().getOperationId(), operation.getOperationId());
-        }
+    private void attach(ProgressOperation operation, StyledLabel label) {
+        AssociationLabel association = new AssociationLabel(operation, label);
+        operationIdToAssignedLabels.put(operation.getOperationId(), association);
+    }
 
+    private void detach(ProgressOperation operation) {
         if (!isRenderable(operation)) {
             return;
         }
 
-        AssociationLabel association = operationIdToAssignedLabels.remove(operation.getOperationId());
-        if (association != null) {
-            unusedProgressLabels.push(association.label);
-            if (operation.getParent() != null) {
-                attach(operation.getParent());
-            } else if (!unassignedProgressOperations.isEmpty()){
-                attach(unassignedProgressOperations.pop());
-            }
-        } else {
-            unassignedProgressOperations.remove(operation);
-        }
-    }
-
-    private void addDirectChildOperationId(OperationIdentifier parentId, OperationIdentifier childId) {
-        Set<OperationIdentifier> children = parentIdToChildrenIds.get(parentId);
-        if (children == null) {
-            children = new HashSet<OperationIdentifier>();
-            parentIdToChildrenIds.put(parentId, children);
-        }
-        children.add(childId);
-    }
+        detach(operation.getOperationId());
+        unassignedProgressOperations.remove(operation);
 
-    private void removeDirectChildOperationId(OperationIdentifier parentId, OperationIdentifier childId) {
-        Set<OperationIdentifier> children = parentIdToChildrenIds.get(parentId);
-        if (children == null) {
-            return;
-        }
-        children.remove(childId);
-        if (children.isEmpty()) {
-            parentIdToChildrenIds.remove(parentId);
+        if (operation.getParent() != null) {
+            attach(operation.getParent());
+        } else if (!unassignedProgressOperations.isEmpty()) {
+            attach(unassignedProgressOperations.pop());
```

Comment:
Here we shouldn't have any non-renderable item in that list as we ignore any non-renderable operation in `attach`